staging/src/k8s.io/client-go/discovery/cached/memcache.go

```diff
@@ -96,18 +98,24 @@ func (d *memCacheClient) RESTClient() restclient.Interface {
 	return d.delegate.RESTClient()
 }
 
-// TODO: Should this also be cached? The results seem more likely to be
-// inconsistent with ServerGroups and ServerResources given the requirement to
-// actively Invalidate.
 func (d *memCacheClient) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
-	return d.delegate.ServerPreferredResources()
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	if !d.cacheValid {
+		// If not cached, returns from the delegate.
+		return d.delegate.ServerPreferredResources()
```

Comment:
The other methods return `ErrCacheEmpty` in this case. Why do we call the delegate here?

Reply:
You're right. It should return `ErrCacheEmpty` to have the same logic with other methods. I will update soon.

Reply:
Updated, PTAL @sttts